diff --git a/sway/input/keyboard.c b/sway/input/keyboard.c
index c5a646c..2ae18b5 100644
--- a/sway/input/keyboard.c
+++ b/sway/input/keyboard.c
@@ -675,6 +675,10 @@ static void handle_modifier_event(struct sway_keyboard *keyboard) {
 		if (!wlr_keyboard_group_from_wlr_keyboard(keyboard->wlr)) {
 			ipc_event_input("xkb_layout", keyboard->seat_device->input_device);
 		}
+	} else {
+	  if (!wlr_keyboard_group_from_wlr_keyboard(keyboard->wlr)) {
+	        ipc_event_input("mod_change", keyboard->seat_device->input_device);
+	  }
 	}
 }
 
diff --git a/sway/ipc-json.c b/sway/ipc-json.c
index 5f6794c..79e648f 100644
--- a/sway/ipc-json.c
+++ b/sway/ipc-json.c
@@ -1065,6 +1065,8 @@ json_object *ipc_json_describe_input(struct sway_input_device *device) {
 
 		xkb_layout_index_t num_layouts = xkb_keymap_num_layouts(keymap);
 		xkb_layout_index_t layout_idx;
+		uint32_t pressed_mods_bitmask = wlr_keyboard_get_modifiers(keyboard);
+		json_object_object_add(object, "mods_mask", json_object_new_int(pressed_mods_bitmask));
 		for (layout_idx = 0; layout_idx < num_layouts; layout_idx++) {
 			const char *layout = xkb_keymap_layout_get_name(keymap, layout_idx);
 			json_object_array_add(layouts_arr,
